File: WiFiConnect/app/src/main/java/com/lilanz/wificonnect/utils/SerialPortUtil.java
package com.lilanz.wificonnect.utils;

public class SerialPortUtil {

    /**
     * @param buffer 串口传回来的16进制数据
     * @return 屏幕能够显示的16进制数据
     */
    public static String byte2hexStr(byte[] buffer) {
        String h = "";
        for (int i = 0; i < buffer.length; i++) {
            String temp = Integer.toHexString(buffer[i] & 0xFF);
            if (temp.length() == 1) {
                temp = "0" + temp;
            }
            h = h + temp;
        }

        return h;
    }

    /**
     * @param inputStr 屏幕端显示的16进制字符串
     * @return 机器能够识别的16进制字节（可以直接发给串口使用）
     */
    public static byte[] hexStr2Byte(String inputStr) {
        if (inputStr == null || inputStr.equals("")) {
            return null;
        }
        inputStr = inputStr.replace(" ", "");
        byte[] baKeyword = new byte[inputStr.length() / 2];
        for (int i = 0; i < baKeyword.length; i++) {
            try {
                baKeyword[i] = (byte) (0xff & Integer.parseInt(inputStr.substring(i * 2, i * 2 + 2), 16));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return baKeyword;
    }



}

File: KotlinTool/app/src/main/java/com/lilanz/kotlintool/utils/SerialPortUtil.kt
package com.lilanz.kotlintool.utils

import kotlin.experimental.and

/**
 * 都是机器数据与屏幕显示数据间的转换
 */
class SerialPortUtil {
    companion object {
        /**
         * 串口16进制数据 转 屏幕显示16进制数据
         */
        fun byte2HexStr(buffer: ByteArray): String {
            var str = ""
            for (byte in buffer) {
                var temp = Integer.toHexString((byte and 0xFF.toByte()).toInt())
                if (temp.length == 1) {
                    temp = "0" + temp
                } else if (temp.length > 2) {
                    temp = temp.substring(temp.length - 2, temp.length)
                }
                str += temp
            }
            return str
        }

        /**
         * 16进制屏幕数据 转 16进制串口数据
         */
        fun hexStr2Byte(str: String?): ByteArray? {
            if (str != null && !"".equals(str)) {
                var temp = str.replace(" ", "")
                var byteArray = ByteArray(temp.length / 2)

                for (i in 0 until byteArray.size) {
                    byteArray[i] = (0xff and temp.substring(i * 2, i * 2 + 2).toInt(16)).toByte()
                }
                return byteArray
            }
            return null
        }
    }
}